### PR TITLE
Enable tests and update features to support new language changes around patterns.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**

### CODE REVIEW

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 1]
```diff
@@ -193,33 +193,33 @@ public async Task TestForCatchClause_NotAfterFilter3()
         public async Task TestForSwitchCase_NotInEmptySwitchStatement() =>
             await VerifyAbsenceAsync(AddInsideMethod(@"switch (1) { $$ }"));
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestForSwitchCase_SemanticCheck_NotAfterPredefinedType() =>
-            await VerifyAbsenceAsync(AddInsideMethod(@"switch (new object()) { case int $$ }"));
+            await VerifyKeywordAsync(AddInsideMethod(@"switch (new object()) { case int $$ }"));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs:198]: this is the newly allowed behavior.  i.e. you can say `case int when ...` instead of `case int i when ...`

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 2]
```diff
@@ -462,19 +462,19 @@ class C
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestForSwitchCase_SemanticCheck_AfterLocalConstantVar() =>
-            await VerifyKeywordAsync(AddInsideMethod(@"const object var = null; switch (new object()) { case var $$ }"));
+            await VerifyAbsenceAsync(AddInsideMethod(@"const object var = null; switch (new object()) { case var $$ }"));
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs:465]: i specifically regressed this behavior because:
1. teh logic to support it was such a pain to keep
2. supporting users explicitly namign things like `var` is explicitly carved out as something we do not cater for in the IDE.

File: src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs [hunk 3]
```diff
@@ -22,20 +22,14 @@ public WhenKeywordRecommender()
         protected override bool IsValidContext(int position, CSharpSyntaxContext context, CancellationToken cancellationToken)
         {
             return context.IsCatchFilterContext ||
-                (IsAfterCompleteExpressionOrPatternInCaseLabel(context, out var expressionOrPattern) &&
-                !IsTypeName(expressionOrPattern, context.SemanticModel, cancellationToken));
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/Completion/KeywordRecommenders/WhenKeywordRecommender.cs:26]: all the logic removed was just jumping through hoops to try to prevent `when` from showing up too much *and* to support `case var when`.  i.e. having `var` not be a variable-pattern, but rather some sort of constant pattern.  It was intefering with supporting real cases simply and just added too much complexity for a scenario we do not consider important. So it's all been removed and replaced with something much simpler and easier to maintain.

File: src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs [hunk 4]
```diff
@@ -503,7 +503,7 @@ class D
     public int P2 { get; set; }
 }
 ";
-            await VerifyNoItemsExistAsync(markup);
+            await VerifyItemExistsAsync(markup, "P1");
```
Comment (allisonchou) [inline src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs:506]: Just trying to better understand, why was this file changed? Does it have to do with the `when` keyword or is it something separate?

Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs:506]: this used to be illegal in C# but became legal in C# 9.0. i.e. you can continually nest recursive patterns within each other :)

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 5]
```diff
@@ -193,33 +193,33 @@ public async Task TestForCatchClause_NotAfterFilter3()
...
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestForSwitchCase_SemanticCheck_NotAfterPredefinedType() =>
-            await VerifyAbsenceAsync(AddInsideMethod(@"switch (new object()) { case int $$ }"));
+            await VerifyKeywordAsync(AddInsideMethod(@"switch (new object()) { case int $$ }"));
 
-        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/40015"), Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
+        [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestForSwitchCase_SemanticCheck_NotAfterPredefinedType_BeforeBreak() =>
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs:201]: Update the names of the tests to no longer say "Not" when we expect it

File: src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs [hunk 6]
```diff
@@ -502,7 +502,7 @@ class C
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
         public async Task TestForSwitchCase_SemanticCheck_AfterTypeAliasAndFieldConstantVar()
         {
-            await VerifyKeywordAsync(@"
+            await VerifyAbsenceAsync(@"
```
Comment (ryzngard) [inline src/EditorFeatures/CSharpTest2/Recommendations/WhenKeywordRecommenderTests.cs:505]: Update test names to have "Not"
